sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
         # Approximation of x * gaussian_cdf(x) used by Hendrycks and Gimpel
         return 0.5 * data * (1 + mx.sym.Activation((math.sqrt(2 / math.pi) * (data + (0.044715 * (data**3)))),
                                                    act_type="tanh"))
+    elif act_type == C.NO_ACTIVATION:
+        return data
     else:
         return mx.sym.Activation(data, act_type=act_type)
 
 
-class LayerNormalization:
+class Layer(ABC):
+
+    def get_max_seq_len(self) -> Optional[int]:
+        """
+        :return: The maximum length supported by the layer if such a restriction exists.
+        """
+        return None
+
+    @abstractmethod
+    def get_num_hidden(self) -> int:
+        """
+        :return: The representation size of this layer.
+        """
+        raise NotImplementedError()
+
+
+class EncoderLayer(Layer):
+    """
+    Generic encoder layer interface for a layer which takes the previous hidden state and sequence lengths and produces
```

Comment:
I like the description but it sounds a little bit like a stepwise application, whereas `encode_sequence` runs over a whole sequence. Maybe we can rephrase that it takes a sequence of certain representation size, and encodes it to a new representation for each time-step but it does that potentially in parallel (transformer)?

Reply:
oh, I guess "previous hidden state" is ambiguous. I meant the hidden state(s!) of the previous layer. I will try to make it clearer that that is what is meant.

Reply:
how is "Generic encoder layer interface for a layer which takes the sequence of previous hidden states and sequence lengths and produces a new sequence hidden states, potentially changing the sequence length."?